### ISSUE TITLE
Integrating Barcode Buddy into grocy-desktop 

### ISSUE THREAD

**Issue by Forceu** [grocy/grocy-desktop#13]
Created from this reddit thread: https://www.reddit.com/r/grocy/comments/fr4i5z/grocy_desktop_win10_wwebcam_german_barcode/flu46jm/

> I personally haven't looked at Barcode Buddy in detail yet (shame on me), but I'm looking forward to integrate it in grocy-desktop (the Open Food Facts lookup is a thing many users request) - as it's also a PHP app, it should work, I think.
> 
> So just a rough brain storming:
> 
> GUI: Maybe create tabs for grocy and Barcode Buddy
> Start a separate PHP process for Barcode Buddy
> Auto configure a grocy API key for Barcode Buddy
> Make Barcode Buddy (and therefore grocy) available from outside the machine (automatically care about firewall config)
> Does Barcode Buddy only communicate via the API?
> 
> Barcode scanning: We maybe could add a native textbox or something in the application to grab barcodes - is the mobile view (from your README screenshot) just a web page Barcode Buddy provides or something else?
> 
> (Maybe we should continue this in an issue.)

Great to hear that you would consider integrating it!

* A seperate tab would be probably the best choice
* I could add a config variable, so if Barcode Buddy is in the grocy-desktop, it would skip the setup and use the details provided by Grocy
* Barcode Buddy only communicates with the Grocy API and OpenFoodFacts API
* Barocode Scanning: There are different ways to provide barcodes, either by entering them in a textbox in the UI or through curl / internal PHP requests (I am adding a proper API soon).

**Comment by berrnd** [comment]
> I could add a config variable

If you mean environment variables, that would be great. I saw that it's just a POST request to `setup.php` - would also be possible to do just that request automatically.

I noticed that you also use a `data` directory - would be good to have this separate from the application directory via maybe also an environment variable (makes updates easier, and on Windows it's common to have (user) applications in `%localappdata%` and their data in `%appdata%`).

In general, maybe the same approach as grocy uses would be good for configuration options: Any option can be overriden by an environment variable.
([grocy code ref](https://github.com/grocy/grocy/blob/f092c8b10d7f9d4a7e275b38534d78c8ecce0817/helpers/extensions.php#L155-L174) - the thing about "settingoverrides"-txt-files should be removed soon, that's currently used in grocy-desktop.)

I'll try to do some first tests in the coming week.

**Comment by Forceu** [comment]
Yes, you are absolutely right, you could simply setup the app with a POST request. 

The database location can be set in `incl/config.php` and at the moment only the database is saved by default to the `data` folder.

Setting all available options with environment variables would a good way as well, I will add it to my todo list

**Comment by Forceu** [comment]
User config can now be overridden in config.php file https://github.com/Forceu/barcodebuddy/commit/8cc0630c3b6c91bdf3de57cfbdfc38e1280529a5

**Comment by berrnd** [comment]
I fiddled around a little bit and it basically works. :)

Set that little screenvideo: https://nc.b7c7.de/s/3rqKQPEd7QCC2nc

But it would be really good to have environment variables for all `incl/config.php` settings - it's kind of hard/ugly to write/modify a PHP file from C# code automatically...

And I need to find a proper way to download the latest release ZIP during the build process & for in-app updates (currently I used the master branch)...

**Comment by Forceu** [comment]
Looks great! Yes, environment variables would be probably the best way for that.

How do you update the included grocy instance? I could probably adapt that way for updating Barcode Buddy

**Comment by berrnd** [comment]
> How do you update the included grocy instance?

I host a fixed path which always returns the latest relesae ZIP (https://releases.grocy.info/latest).

But I already implemented that now by using the GitHub API to get your latest tagged archive - works so far...

https://github.com/grocy/grocy-desktop/blob/f0f26883886a040877649c83ae8877a42600e966/grocy-desktop/GrocyDesktopDependencyManager.cs#L15-L27

**Comment by Forceu** [comment]
Environment variables are added in  https://github.com/Forceu/barcodebuddy/commit/f8d8f7ad30ba71dc8e075c036a9395a52e1b93d2
and https://github.com/Forceu/barcodebuddy/commit/09046c31f61cf744f9152e2c8ded22dac3d6c005

_I renamed `OVERRIDDEN_CONFIG` to `OVERRIDDEN_USER_CONFIG`._ 

To override any constant declared in `incl/config.php`, simply export the environment variable `BBUDDY_X` and replace the X. Eg. to set `CURL_TIMEOUT_S`, the environment variable has to be named `BBUDDY_CURL_TIMEOUT_S`.

The constant `OVERRIDDEN_USER_CONFIG` stays an array, to pass arrays use the delimiter `;`, eg. `BBUDDY_OVERRIDDEN_USER_CONFIG="GROCY_API_URL=https://myurl/api/;GROCY_API_KEY=1234"`

**Comment by berrnd** [comment]
Great, so it's basically ready (currently using your master branch).

One thing I noticed:
Barcode Buddy does not skip the inital setup page if the required `GROCY_API_URL` and `GROCY_API_KEY` are already set through environment variables.

Open is if and how to make it possible to access Barcode Buddy from outside the machine / grocy-desktop.

If you want to test it, here is the current .msi install package: https://nc.b7c7.de/s/mkwjSNbwddwN8A2
(Can be cleanly removed by uninstalling it, only creates files in `%appdata%\grocy-desktop` and `%localappdata%\grocy-desktop`.)

Barcode Buddy is disabled by default, can be enabled under File -> Enable Barcode Buddy.

**Comment by Forceu** [comment]
Weird, so I tested it with a clean install with the current master, and if both values are set, it skips right to the main ui. I think in the commit added above, you are not using the array for the user config, but creating two env variables for the API URL and API key. For these you need to set the env variable like this:

`BBUDDY_OVERRIDDEN_USER_CONFIG="GROCY_API_URL=https://myurl/api/;GROCY_API_KEY=1234"`


To be honest I guess that is not the best way to do it, I might consider scrapping the array thing for the user config.

**Comment by berrnd** [comment]
> but creating two env variables for the API URL and API key

That's correct, sorry, seems that I haven't read your last comment to the end... :D

So can of course also be set in that way...

**Comment by Forceu** [comment]
For your information, I added an authentication module in the current master, you can disable it with the environment variable` BBUDDY_DISABLE_AUTHENTICATION=true`. Also the config.php was moved to /config-dist.php, but I suppose you are not modifying that one anymore?

**Comment by berrnd** [comment]
Thanks, will adjust that. The `data/config.php` is maybe not needed for grocy-desktop if all required settings will be set via env vars. Just if the user wants to manually change something there...

So another good thing to have would be that the `data` directory location would be also changeable through an env var (would be good to have it outside the appllication directory for updates, so that this can just be replaced without keeping anything - also do it that way for grocy...).

**Comment by Forceu** [comment]
Added in https://github.com/Forceu/barcodebuddy/commit/8c4ca9859a117912b4dafe57585c5ab3c05498e7

I prefer not to make a variable for the data directory, but rather three different ones for the files that are included inside.

env `BBUDDY_CONFIG_PATH` is the path including filename where config.php is located. Default: `/data/config.php`

env `BBUDDY_AUTHDB_PATH` is the path including filename where the database containing username / hashed password is located. Default: `/data/users.db`

env `BBUDDY_DATABASE_PATH` is the path including filename where the database containing all user settings is located. Default: `/data/barcodebuddy.db`

**Comment by berrnd** [comment]
Thanks, just did the changes.

Now the setup screen seems not to be skipped again.
Environment variables are set like this: 
![grafik](https://user-images.githubusercontent.com/2161815/78602331-932de300-7856-11ea-8dfc-871cd7caeff2.png)

Here is a new build if you want to test it: https://nc.b7c7.de/s/tqsGMD4Z2izqsBA

**Comment by Forceu** [comment]
Still the same problem as above ;-)

The user config needs to be in one variable like this:

`BBUDDY_OVERRIDDEN_USER_CONFIG` = `"GROCY_API_URL=https://myurl/api/;GROCY_API_KEY=1234"`

All other variables are correct I think

**Comment by berrnd** [comment]
Oh, I'm so sorry to have missed that again... :/

Is now fixed by https://github.com/grocy/grocy-desktop/commit/037576d58f59b4816b8425faa3a8dfc8a73ae638 and works ... almost - the "Create User" screen is not skipped (`DISABLE_AUTHENTICATION` is set).

`BBUDDY_OVERRIDDEN_USER_CONFIG` is set to:
```
GROCY_API_URL=http://localhost:56624/api/;GROCY_API_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx;DISABLE_AUTHENTICATION=true;CONFIG_PATH=C:\Users\bb\AppData\Roaming\grocy-desktop\barcodebuddy-data\config.php;AUTHDB_PATH=C:\Users\bb\AppData\Roaming\grocy-desktop\barcodebuddy-data\users.db;DATABASE_PATH=C:\Users\bb\AppData\Roaming\grocy-desktop\barcodebuddy-data\barcodebuddy.db
```

**Comment by Forceu** [comment]
https://github.com/grocy/grocy-desktop/pull/15

Sorry, I know it is confusing with the array and I still haven't written the documentation for it. Basically only the values that the user can change in the UI go into the array, all other don't. I created a pull request to fix it

**Comment by berrnd** [comment]
So I think this is now read. :)

Further things I did:
- Use the latest Barcode Buddy release instead of the master branch
- External access is now possible for both grocy and Barcode Buddy (can be enabled in the file menu, URL info is displayed then at the bottom of the window)
  - The app restarts and the native Windows Firewall dialog pops up, which need to be accepted accordingly
- There is now a fixed port used (`4010` for grocy, `4011` for Barcode Buddy) unless those ports are already used

Here is onther build if you want to try it: https://nc.b7c7.de/s/3FdXLqsXdzaKiBT

(Will fix the other two open issues and then drop a new release in the next days.)

**Comment by Forceu** [comment]
Looks great, thank you very much for your hard work!

A couple of things I noticed however (tested on a Windows 10 machine):

- When looking up an unknown barcode, openfoodfacts SSL certificate is not accepted (probably because it is signed by Letsencrypt). I don't know if you can do anything about it, if not, this issue can be solved by ignoring invalid SSL certificates (env `CURL_ALLOW_INSECURE_SSL_CA` = `true`)
- A config.php is created in the /data/ folder and the regular barcodebuddy-data folder. Could be a bug in my code, however it seems that the one in the /data/ folder is being used
- The websocket server is not started, is it possible to run it as well? It can be started by running `php  wsserver.php` in a background process.

Great work :)

**Comment by berrnd** [comment]
Thanks for testing!

> - When looking up an unknown barcode, openfoodfacts SSL certificate is not accepted (probably because it is signed by Letsencrypt). I don't know if you can do anything about it, if not, this issue can be solved by ignoring invalid SSL certificates (env `CURL_ALLOW_INSECURE_SSL_CA` = `true`)

I have added to set this environment variable.

> - A config.php is created in the /data/ folder and the regular barcodebuddy-data folder. Could be a bug in my code, however it seems that the one in the /data/ folder is being used

Seems not to be done by grocy-desktop, but I think that everything needed is in the `barcodebuddy-data` folder (which will also backed up/restored when using the backup/restore function).

> - The websocket server is not started, is it possible to run it as well? It can be started by running `php  wsserver.php` in a background process.

Also added that, but now somehow the `screen.php` page is kind of (really) slow. Any idea? The output of all (3) PHP processes can be shown via Help/Debug/Show PHP Process Output.

New build: https://nc.b7c7.de/s/PMsA5JbMcNfDWM2

**Comment by Forceu** [comment]
I just checked my code, I had a bug that the default config.php was loaded before environment variables were checked. Fixed in https://github.com/Forceu/barcodebuddy/commit/fba041dda40b1771c49fd5d1a32caf128fcc49fd.

I also added the environment variable `HIDE_LINK_GROCY`, so the link in the header can be hidden (if set to `true`).

I tried your new build, unfortunately the websocket server does not seem to start? I installed it over the old one. By clicking on Barcode Buddy / Settings, the websocket status tab says it is not running and the PHP process output only shows two processes (Grocy and Barcode Buddy).

Regarding the performance: This could be because the server does not allocate enough child processes. At the moment using the websocket server spawns an additional process per client unfortunately. So if both the menu and the screen module are active, there would be a total of three running PHP processes including the wsserver.php:

After rendering the web ui, a javascript connects to `/incl/ssesse_data.php`, which is an internal PHP websocket client running for up to 120s, sending updates to the JS client. Another one is spawned for the screen module.

In the beginning I had a JS websocket client which was very efficient, unfortunately a lot of users had problems configuring it. That is the reason why I went to this model, as it does not require any configuration (but uses _a lot_ more resources).

Are there any ways to increase the number of child processes? That would probably solve the problem.

**Comment by berrnd** [comment]
> After rendering the web ui, a javascript connects to `/incl/ssesse_data.php`, which is an internal PHP websocket client running for up to 120s, sending updates to the JS client.

That's then the problem: The PHP built-in development webserver is only single threaded, so can't handle more than one request in parallel: https://www.php.net/manual/en/features.commandline.webserver.php

I think that's not solvable, unless using another/real webserver...

**Comment by Forceu** [comment]
Oh I see. In that case I will probably add an option to use JS client again in the next version. For the meantime it would be probably best to simply hide the Screen module and not start the websocketserver (although that will add lag to Barcode Buddy, as it will be trying to connect).

**Comment by berrnd** [comment]
Or maybe we just use and include nginx (runs also on Windows), because why not... 😀

**Comment by Forceu** [comment]
That would probably increase the performance of Grocy as well! Is there any way I can help? Unfortunately I am almost exclusively using Linux, therefore I don't have any .NET/C# experience

**Comment by berrnd** [comment]
Should be no big difference, it's also only starting a process - just the nginx configuration (file) needs to be somehow generated automatically based on the configured ports/paths/etc.

Will try it later...

> I tried your new build, unfortunately the websocket server does not seem to start? I installed it over the old one. By clicking on Barcode Buddy / Settings, the websocket status tab says it is not running and the PHP process output only shows two processes (Grocy and Barcode Buddy).

For that: I have not changed the version number between the test-builds, so make sure to always uninstall and delete `%localappdata%\grocy-desktop` and `%appdata%\grocy-desktop` (the first should be deleted by the uninstaller, the second not as it contains user data) before installing a new .msi.

**Comment by berrnd** [comment]
nginx works (so far). :)

New test build: 

Two problems: https://nc.b7c7.de/s/fbDrCNsf8SkCgpE
- Barcode Buddys `screen.php` is still slow, however
- Not all nginx processes get properly closed on application exit (therefore the default port(s) are already in use when restarting grocy-desktop)

**Comment by berrnd** [comment]
> Two problems: 
> - Barcode Buddys `screen.php` is still slow, however
> - Not all nginx processes get properly closed on application exit (therefore the default port(s) are already in use when restarting grocy-desktop)

Both is now fixed. :)
Will push the new release tomorrow.

Current build, if you want to give it a last try: https://nc.b7c7.de/s/Nj7AcXs4czkGX3k

**Comment by Forceu** [comment]
Alright so I just tried it, and unfortunately it only displays a blank page. The problem was probably a space character in my Windows username, therefore the root folder had a space as well.

To fix, you can use single quotes for the root argument in line 44 (root `'C:\test folder\'`). 

Also for some reason it was complaining that localhost was invalid, maybe change it to 127.0.0.1? Although it does kind of surprise me.

Unfortunately I could not try it myself, as the nginx.conf file was always overwritten.

**Comment by berrnd** [comment]
Thanks for testing and noticing that.

I think both should be fixed, new build: https://nc.b7c7.de/s/XioHiME9FnXym5b

(You can edit the file `%appdata%\grocy-desktop\runtime-dependencies\1.6.0.0\nginx\conf\nginx.conf.template` if you want to tweak the nginx configuration manually, that file get copied and the `$xx$` placeholders replaced.)

**Comment by Forceu** [comment]
It is actually showing up now :)

One more thing: I think you are using an incorrect nginx conf for BarcodeBuddy - you can find an example here: https://github.com/Forceu/barcodebuddy/blob/master/example/nginxConfiguration.conf

You would need to change

```
	server {
		listen $INTERFACE$:$BARCODEBUDDYPORT$;

		root "$BARCODEBUDDYROOT$";

		location / {
			try_files $uri /index.php$is_args$query_string;

			location ~ \.php$ {
				fastcgi_pass 127.0.0.1:$PHPFASTCGIPORT$;
				fastcgi_index index.php;
				fastcgi_intercept_errors on;
				fastcgi_buffer_size 16k;
				fastcgi_buffers 4 16k;
				fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
				include fastcgi_params;
			}
		}
	}
```


to 

```
	server {
		listen $INTERFACE$:$BARCODEBUDDYPORT$;

		root "$BARCODEBUDDYROOT$";
		index index.php index.html index.htm;

		client_max_body_size 20M;
		client_body_buffer_size 128k;


		location / {
			try_files $uri $uri/ =404;
		}

		location /api/ {
			try_files $uri /api/index.php$is_args$query_string;
		}

		location ~ /data/ {
		        deny all;
		}

		location ~ \.php$ {
			fastcgi_pass 127.0.0.1:$PHPFASTCGIPORT$;
			fastcgi_index index.php;
			fastcgi_intercept_errors on;
			fastcgi_buffer_size 16k;
			fastcgi_buffers 4 16k;
			fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
			include fastcgi_params;
		}
	}
```

Also changing `worker_processes 2;` to `worker_processes auto;` improves performance.


_However_ it seems that the underlying php engine is single threaded?? Nginx gives a gateway timeout, once the JavaScript is trying to connect to the SSE php file.

**Comment by berrnd** [comment]
I've added your suggested tweaks, thanks.

> However it seems that the underlying php engine is single threaded?? Nginx gives a gateway timeout, once the JavaScript is trying to connect to the SSE php file.

Seems so. Since there is no PHP-FPM for Windows, the recommended solution seems to be just to spin up some more `php-cgi.exe` processes and use them in a nginx upstream server pool - I added that also with 8 processes, seems to be snappier now.

Next build: https://nc.b7c7.de/s/5QnNHeogtDzTLK3

**Comment by Forceu** [comment]
Yes, now everything is working! Great work :)

**Comment by berrnd** [comment]
Great! Also thanks again for your help on this - and even more for your ongoing work on Barcode Buddy itself! :)

So I published the new release, let's see how it goes...

https://www.reddit.com/r/grocy/comments/g2lttr/grocydesktop_has_now_barcode_buddy_integraded_and/
